src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
-	opts bootstrap.RunOptions,
-) (result.IndexBootstrapResult, error) {
-	if shardsTimeRanges.IsEmpty() {
-		return result.NewIndexBootstrapResult(), nil
-	}
-	step := newBootstrapIndexStep(namespace, b.src, b.next, opts)
-	err := b.runBootstrapStep(namespace, shardsTimeRanges, step)
-	if err != nil {
-		return nil, err
+	curr := bootstrap.Namespaces{
+		Namespaces: bootstrap.NewNamespacesMap(bootstrap.NamespacesMapOptions{}),
 	}
-	return step.result(), nil
-}
+	for _, elem := range namespaces.Namespaces.Iter() {
+		id := elem.Key()
+		namespace := elem.Value()
 
-func (b baseBootstrapper) runBootstrapStep(
-	namespace namespace.Metadata,
-	totalRanges result.ShardTimeRanges,
-	step bootstrapStep,
-) error {
-	prepareResult, err := step.prepare(totalRanges)
-	if err != nil {
-		return err
-	}
+		// Shallow copy the namespace, do not modify namespaces input to bootstrap call.
+		currNamespace := namespace
 
-	var (
-		wg                     sync.WaitGroup
-		currStatus, nextStatus bootstrapStepStatus
-		currErr, nextErr       error
-		nextAttempted          bool
-	)
-	currRanges := prepareResult.currAvailable
-	nextRanges := totalRanges.Copy()
-	nextRanges.Subtract(currRanges)
-	if !nextRanges.IsEmpty() &&
-		b.Can(bootstrap.BootstrapParallel) &&
-		b.next.Can(bootstrap.BootstrapParallel) {
-		// If ranges can be bootstrapped now from the next source then begin attempt now
-		nextAttempted = true
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			nextStatus, nextErr = step.runNextStep(nextRanges)
-		}()
-	}
+		dataRange := currNamespace.DataRunOptions.ShardTimeRanges.MinMaxRange()
+		requestedLogFields := append(logFieldsCopy(logFields), []zapcore.Field{
+			zap.String("namespace", id.String()),
+			zap.Int("numShards", len(currNamespace.Shards)),
+			zap.Duration("dataRange", dataRange),
+		}...)
+		if dataRange > 0 {
+			dataMin, dataMax := currNamespace.DataRunOptions.ShardTimeRanges.MinMax()
+			requestedLogFields = append(requestedLogFields, []zapcore.Field{
+				zap.Time("dataFrom", dataMin),
+				zap.Time("dataTo", dataMax),
+			}...)
+		}
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			indexRange := currNamespace.IndexRunOptions.ShardTimeRanges.MinMaxRange()
+			requestedLogFields = append(requestedLogFields, []zapcore.Field{
+				zap.Duration("indexRange", indexRange),
+			}...)
+			if indexRange > 0 {
+				indexMin, indexMax := currNamespace.IndexRunOptions.ShardTimeRanges.MinMax()
+				requestedLogFields = append(requestedLogFields, []zapcore.Field{
+					zap.Time("indexFrom", indexMin),
+					zap.Time("indexTo", indexMax),
+				}...)
+			}
+		}
 
-	min, max := currRanges.MinMax()
-	logFields := []zapcore.Field{
-		zap.String("source", b.name),
-		zap.String("namespace", namespace.ID().String()),
-		zap.Time("from", min),
-		zap.Time("to", max),
-		zap.Duration("range", max.Sub(min)),
-		zap.Int("shards", len(currRanges)),
+		b.log.Info("bootstrap from source for namespace requested", requestedLogFields...)
```

Comment:
nit: maybe `"bootstraping from source"`? It should have the namespace as a field already so distinction not super neccessary